zip-0311.rst

```diff
@@ -0,0 +1,111 @@
+==========
+FlyClient reduces the number of block headers needed for light client verification of a valid chain, from linear (as in the current reference protocol) to logarithmic in blockchain length. This verification is correct with high probability. It also allows creation of subchain proofs, so light clients need only check blocks later than the most recently verified block index. Following that, verification of a transaction inclusion within that block follows the usual reference protocol [#ZIPXXX]_. 
+
+A smaller proof size could facilitate ‘side-chaining’ with other blockchains such as Ethereum. It also reduces bandwidth and storage requirements for limited clients like mobile or IoT devices.
+
+
+Specification
+=============
+
+Probabilistic sampling
+```````````````````````
+Instead of downloading every block header, the FlyClient samples blocks from the full node with the given sampling distribution. Here, x is the relative aggregate weight at the given block (e.g. if x=1/2 at some block, it means that half the chain's difficulty has been amassed in the interval up to that block), and δ is the fraction of the blocks which are always checked by the client with probability 1.
+
+.. figure:: pdf.png
+    :align: center
+    :scale: 50 %
+    :figclass: align-center
+    
+It has been found that as long as log_c(δ) is constant, where c is the adversary's mining power as a fraction of the honest mining power, then the number of light client queries needed are logarithmic in the size of the blockchain [#FlyClient]_.
+
+
+New block header
+`````````````````
+Each FlyClient prover has to maintain a commitment to all its block headers in a Merkle mountain range (MMR). This means that each ZCash block header must include the MMR root of all the blocks before it. This can be implemented in a hard fork of the ZCash protocol, in which the MMR root is added to the header of all old and new blocks. 
+
+Merkle mountain range
+``````````````````````
+
+.. figure:: MMR.png
+    :align: center
+    :figclass: align-center
+
+    Schematic of Merkle mountain range (from [#FlyClient]_).
+    
+The MMR structure allows for subchain proofs. Consider a client who has already received a verified a proof for a chain of n blocks. When it later receives k new blocks, it need only verify that the new section of length k is honest, as well as that the old chain is a prefix of the new chain.
+
+.. figure:: MMRWeight.png
+    :align: center
+    :figclass: align-center
+
+Besides the block hash (or the hash of its children nodes), each MMR node contains an additional field of 8 bytes to store the cumulative difficulty of the nodes below it.
+
+Verifier
+`````````
+**Verifying a valid chain**: the hash of the block B_n at the head of the chain specifies a random subset of blocks to sample. If the PoW solution or the MMR proofs of any of the sampled blocks is invalid, then the verifier rejects the proof.  Otherwise, it accepts B_n as the last block of the honest chain.
```

Comment:
This would need to specify the sampled subset. Again, it's enough for a draft.

Reply:
Specifying the consensus rule changes is IMO more important than specifying the concrete light client strategy. You certainly need to convince readers that a good light client strategy that is enabled by the consensus rule changes exists, but it's not as important that an exact specific one is fully nailed down, that's a discussion that can easily be done even after the network upgrade.